### DISCUSSION TITLE
After livewire request, AlpineJS variable defined by Livewire property doesn't update

### DISCUSSION THREAD

**Discussion by joostvanhoof** [livewire/livewire#3284] [1. Help] [answered]
**Livewire component:**

```php
<?php

namespace App\Http\Livewire;

use App\Models\Post;
use Livewire\Component;

class Example extends Component
{
    public $postId = null;

    public function openPost($post_id)
    {
        $this->postId = $post_id;
    }

    public function closePost()
    {
        $this->reset('postId');
    }

    public function render()
    {
        return view('livewire.posts-page', [
            'postSingle' => $this->postId ? Post::findOrFail($this->postId) : null,
            'posts' => Post::all()
        ]);
    }
}
```

**`posts-page` Blade component:**
```html
@forelse($posts as $post)
    <div wire:click="openPost({{ $post->id }})"></div>
@empty
    No posts
@endforelse

@if($postSingle)
    @include('post-card')
@endif
```

**`post-card` component:**
```html
 <div
        x-data="{
            priority: {{ $postSingle->priority }}
        }"
        x-init="$nextTick(() => { 
            priority = {{ $postSingle->priority }}
        })"
    >
       <span x-on:click="priority = 1"></span>
       <span x-on:click="priority = 2"></span>

      <span wire:click="closePost"></span>
    </div>
```

This all works as expected. But if you do the following:

1. Click on a post in the loop (the div with `openPost()`
2. It opens the post as expected (within `@if($postSingle)`)
3. Click on a `span` to set the priority
4. Now click on another div in the loop like you did in step 1
5. Now even though `$postSingle` was updated and `x-data` has the correct `priority` value using `$postSingle->priority`, the `priority` Alpine value is still the value from what you clicked on in step 3

Now interestingly it does work as expected if you do:
1. Click on a post in the loop (the div with `openPost()`
2. It opens the post as expected (within `@if($postSingle)`)
3. Click on a `span` to set the `priority` Alpine value
4. Close the post using `wire:click="closePost"`
5. The open another Post, the `priority` value is now correctly set

This second user journey doesn't work properly without the `$nextTick` part, that's why I've added it.

Hopefully this make sense, if you need more info please let me know. I'm guessing something gets mixed up between the Livewire requests but I can not figure out what it is.

Laravel: 8.49.2
Livewire: 2.5.3
AlpineJS: 3.2.1

**Comment by joshhanley** [comment, marked as answer]
@joostvanhoof the reason it won't work, is because Alpine will only parse the x-data the first time it loads, so it won't pick up any changes applied to it. 

The reason closing first works, is because it removes that Alpine component from the DOM and inits a new one the next time.

A potential way to fix it, is on the `div` that contains Alpine `x-data` add a wire:key that us unique to the record being edited. So that way if `$postSingle` changes, Livewire should remove the existing div and add a new one. 

Something like this 
```blade
<div wire:key="post-single-{{ $postSingle->id }}" x-data...
```

Hope this helps!

**Reply by joostvanhoof** [reply]
Thanks a lot! This was indeed the solution. I've been stuck on this for quite some time, might be good to add this to the Livewire docs (outside the troubleshooting section)?

**Reply by joostvanhoof** [reply]
One last question: if I take the functionality to a function outside the component the error is back. Like so:

```javascript
function test() {
  return {
    priority: 0
  }
}
```

and then use `x-data="test()"` (even when using `wire:key`)

**Reply by joostvanhoof** [reply]
@joshhanley just a friendly reminder on the above if you have the time. Would love to do that because it provides more structure to my app as well as syntax highlighting.

**Reply by joshhanley** [reply]
@joostvanhoof sorry, it all looks ok to me. What's it doing?

**Reply by joostvanhoof** [reply]
Don't apologise man, I really appreciate the help and I know you're busy just wanted to send a friendly reminder. I've seen some other people struggle with this so hopefully it helps them as well.

If I refactor `post-card` to this, so take the functionality to a script and also add `wire:key` that we've identified as the solution:

```html
<div x-data="test()" wire:key="{{ $post->id }}">
  <span x-on:click="priority = 1"></span>
  <span x-on:click="priority = 2"></span>

  <span wire:click="closePost"></span>
</div>

<script>
function test() {
  return {
    priority: {{ $postSingle->priority }}
  }
}
</script>
```

Then I run into the original issue as mentioned in point 5 in the first numbered list: even though `$postSingle` was updated and the `test()` function has the correct `priority` value using `$postSingle->priority`, the `priority` Alpine value is still the value from what you clicked on before (so it doesn't update the Alpine value across Livewire requests).

**Reply by joshhanley** [reply]
All good 🙂

So same issue, Alpine won't pick up on the changes to static output once it has been loaded on the screen once.

A better solution is to pass information into the component function using a config object (and then wire:key on the div will work again). It also then ensures the component function is reusable and you could even extract that into a compiled JS bundle if you needed.

So for example

```blade
<div x-data="test({ priority: {{ $postSingle->priority }} })" wire:key="{{ $post->id }}">
  <span x-on:click="priority = 1"></span>
  <span x-on:click="priority = 2"></span>

  <span wire:click="closePost"></span>
</div>

<script>
function test(config) {
  return {
    priority: config.priority
  }
}
</script>
```

**Reply by joostvanhoof** [reply]
Hmm it seems that if I rewrite it to that structure it sometimes works and sometimes doesn't. Sometimes it gives `Unhandled Promise Rejection: ReferenceError: Can't find variable: test` where `test` is the function called in `x-data` and you can see that the component is not correctly initialized. Then when I do a hard page refresh, everything works as expected and this behaviour keeps coming back.

**Reply by joshhanley** [reply]
@joostvanhoof as you using `@push` around the script tag by any chance and not including the Alpine component on the initial render?

**Reply by joostvanhoof** [reply]
It's included on initial render. There is a `@push` but not for Alpine, it's for other functionality.

**Reply by joshhanley** [reply]
@joostvanhoof hmm how odd. If you load your page initially and then open up your browser console and type `test` you should see the function appear. If not, then it's probably not actually loading properly.
